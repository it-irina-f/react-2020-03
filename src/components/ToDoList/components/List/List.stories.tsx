import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, object, array } from "@storybook/addon-knobs";

import { List } from "./List";

export default {
  title: "List Component",
  decorators: [withKnobs],
};

const list = [
  { text: "Полить цветы", isComplete: true },
  { text: "Сделать ДЗ", isComplete: false },
  { text: "Купить продукты", isComplete: true },
  { text: "Приготовить ужин", isComplete: false },
  { text: "Уборка квартиры", isComplete: false },
];

export const componentList = () => [<List key="jsx" list={list} />];
