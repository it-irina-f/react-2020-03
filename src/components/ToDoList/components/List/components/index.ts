export * from "./ListItem";
export * from "./ListItemEdit";
export * from "./Buttons";
