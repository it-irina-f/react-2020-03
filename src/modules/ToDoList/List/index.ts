export * from "./List";
export * from "./ListItemEdit";
