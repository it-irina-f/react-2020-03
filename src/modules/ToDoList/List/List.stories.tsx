import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, object } from "@storybook/addon-knobs";

import { List } from "./List";

export default {
  title: "List Component",
  decorators: [withKnobs],
};

const list = [
  { id: 0, text: "Полить цветы", isComplete: true },
  { id: 1, text: "Сделать ДЗ", isComplete: false },
  { id: 2, text: "Купить продукты", isComplete: true },
  { id: 3, text: "Приготовить ужин", isComplete: false },
  { id: 4, text: "Уборка квартиры", isComplete: false },
];

export const componentList = () => [
  <List
    key="jsx"
    editId={-1}
    list={object("list", list)}
    toggleComplete={action("Item clicked")}
    deleteListItem={action("Click for delete item")}
    handleEdit={action("Click for edit item")}
    saveListItem={action("Click for save editing")}
    filter={"all"}
  />,
];

export const componentListEdit = () => [
  <List
    key="jsx"
    editId={1}
    list={object("list", list)}
    toggleComplete={action("Item clicked")}
    deleteListItem={action("Click for delete item")}
    handleEdit={action("Click for edit item")}
    saveListItem={action("Click for save editing")}
    filter={"all"}
  />,
];
