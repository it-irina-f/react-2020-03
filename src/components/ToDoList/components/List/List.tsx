import React, { FC } from "react";
import styled from "@emotion/styled";
import type { ListItemProps, ToggleCompleteProps } from "types/todo";

import { ListItem } from "./components";

const ListWrapper = styled.div`
  padding: 20px;
  margin-bottom: 20px;
  border: 1px solid #6883bf;
  border-radius: 5px;
`;

interface Props {
  list: ListItemProps[];
  toggleComplete: ToggleCompleteProps;
  deleteListItem: ToggleCompleteProps;
}

interface State {
  list: ListItemProps[];
}

export class List extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      list: this.props.list,
    };
  }

  componentDidMount(): void {
    const newList = this.sortList(this.state.list);
    this.setState({
      list: newList,
    });
  }

  componentDidUpdate(): void {
    const propsList = this.props.list;
    const stateList = this.state.list;

    function isNoChange(propsRow: ListItemProps) {
      return (
        stateList.filter((stateRow) => {
          return (
            stateRow.id === propsRow.id &&
            stateRow.isComplete === propsRow.isComplete
          );
        }).length > 0
      );
    }

    if (
      propsList.every(isNoChange) === false ||
      propsList.length !== stateList.length
    ) {
      const newList = this.sortList(this.props.list);

      this.setState({
        list: newList,
      });
    }
  }

  sortList(initList: ListItemProps[]) {
    const listComplete = initList.filter((row) => {
      return row.isComplete === true;
    });

    const listNoComplete = initList.filter((row) => {
      return row.isComplete === false;
    });

    return listNoComplete.concat(listComplete);
  }

  render() {
    return (
      <ListWrapper>
        {this.state.list.map((row) => (
          <ListItem
            key={row.id}
            listItem={row}
            toggleComplete={this.props.toggleComplete}
            deleteListItem={this.props.deleteListItem}
          />
        ))}
      </ListWrapper>
    );
  }
}
