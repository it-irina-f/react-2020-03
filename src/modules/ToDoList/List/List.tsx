import React, { FC } from "react";
import styled from "@emotion/styled";
import type { ListItemProps, TypeIdNumber, TypeSaveListItem } from "types/todo";
import { ListItem } from "@/components/ListItem";
import { ListItemEdit } from "./ListItemEdit";

const ListWrapper = styled.div`
  padding: 20px;
  margin-bottom: 35px;
  border-radius: 12px;
  box-shadow: rgba(52, 58, 64, 0.15) 0px 1px 10px 0px,
    rgba(52, 58, 64, 0.1) 0px 6px 12px 0px,
    rgba(52, 58, 64, 0.12) 0px 6px 15px -2px;
`;

interface Props {
  list: ListItemProps[];
  toggleComplete: TypeIdNumber;
  deleteListItem: TypeIdNumber;
  handleEdit: TypeIdNumber;
  saveListItem: TypeSaveListItem;
  editId: number;
  filter: string;
}

interface State {
  list: ListItemProps[];
  filter: string;
}

export class List extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      list: this.props.list,
      filter: this.props.filter,
    };
  }

  sortList(initList: ListItemProps[], filter: string) {
    const listComplete = initList.filter((row) => {
      return row.isComplete === true;
    });

    const listNoComplete = initList.filter((row) => {
      return row.isComplete === false;
    });

    switch (filter) {
      case "active":
        return listNoComplete;
        break;
      case "done":
        return listComplete;
        break;
      default:
        return listNoComplete.concat(listComplete);
    }
  }

  render() {
    const sortList = this.sortList(this.props.list, this.props.filter);
    const editId = this.props.editId;
    const itemList = sortList.map((row) =>
      row.id === editId ? (
        <ListItemEdit
          key={row.id}
          listItem={row}
          saveListItem={this.props.saveListItem}
          handleEdit={this.props.handleEdit}
        />
      ) : (
        <ListItem
          key={row.id}
          listItem={row}
          toggleComplete={this.props.toggleComplete}
          deleteListItem={this.props.deleteListItem}
          handleEdit={this.props.handleEdit}
        />
      )
    );
    return (
      <ListWrapper>
        {sortList.length === 0 ? "Список пустой" : itemList}
      </ListWrapper>
    );
  }
}
