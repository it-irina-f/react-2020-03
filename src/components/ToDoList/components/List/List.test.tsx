import React from "react";
import { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";

import { List } from "./List";

const list = [
  { id: 0, text: "Полить цветы", isComplete: true },
  { id: 1, text: "Сделать ДЗ", isComplete: false },
  { id: 2, text: "Купить продукты", isComplete: true },
  { id: 3, text: "Приготовить ужин", isComplete: false },
  { id: 4, text: "Уборка квартиры", isComplete: false },
];

const click = jest.fn();

describe("Component List", () => {
  it("renders list", () => {
    expect(
      renderer
        .create(
          <List
            list={list}
            toggleComplete={click}
            deleteListItem={click}
            editListItem={click}
            cancelEditing={click}
            saveListItem={click}
            editId={-1}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it("render list where 1 editing item", () => {
    expect(
      renderer
        .create(
          <List
            list={list}
            toggleComplete={click}
            deleteListItem={click}
            editListItem={click}
            cancelEditing={click}
            saveListItem={click}
            editId={0}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("simulate onChange checkbox by list", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    wrapper.find("input").at(3).simulate("change");
    expect(click).toHaveBeenCalled();
  });

  it("simulate click Delete List Item", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    wrapper
      .find("div")
      .at(3)
      .find("button[name='deleteListItem']")
      .simulate("click");
    expect(click).toHaveBeenCalled();
  });

  it("check sort list: first element isComplete = false", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    const checkbox = wrapper.find("div").at(1).find("input");
    expect(checkbox.prop("checked")).toBe(false);
  });

  it("check sort list: first element isComplete = true", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    const checkbox = wrapper.find("div").at(5).find("input");
    expect(checkbox.prop("checked")).toBe(true);
  });

  it("check sort list: first element name", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    expect(wrapper.find("div").at(1).find("label").text()).toEqual(
      "Сделать ДЗ"
    );
  });

  it("check sort list: last element name", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    expect(wrapper.find("div").at(5).find("label").text()).toEqual(
      "Купить продукты"
    );
  });

  it("find manage-buttons", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    expect(wrapper.find("button[name='deleteListItem']")).toHaveLength(5);
    expect(wrapper.find("button[name='editListItem']")).toHaveLength(5);
  });

  it("simulate click Edit List Item", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={-1}
      />
    );
    wrapper
      .find("div")
      .at(3)
      .find("button[name='editListItem']")
      .simulate("click");
    expect(click).toHaveBeenCalled();
  });

  it("find form and buttons for editing", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={1}
      />
    );
    expect(wrapper.find("form")).toHaveLength(1);
    expect(wrapper.find("button[name='saveEditing']")).toHaveLength(1);
    expect(wrapper.find("button[name='cancelEditing']")).toHaveLength(1);
  });

  it("simulate click Save List Item", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={1}
      />
    );
    wrapper.find("button[name='saveEditing']").simulate("submit");
    expect(click).toHaveBeenCalled();
  });

  it("simulate click Cancel Editing", () => {
    const wrapper = mount(
      <List
        list={list}
        toggleComplete={click}
        deleteListItem={click}
        editListItem={click}
        cancelEditing={click}
        saveListItem={click}
        editId={1}
      />
    );
    wrapper.find("button[name='cancelEditing']").simulate("click");
    expect(click).toHaveBeenCalled();
  });
});
