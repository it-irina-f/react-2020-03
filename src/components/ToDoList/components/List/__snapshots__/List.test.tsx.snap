// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component List renders list 1`] = `
<div
  className="css-bb4zwd"
>
  <div>
    <label>
      <input
        checked={false}
        name="checkbox_1"
        onChange={[Function]}
        type="checkbox"
      />
      Сделать ДЗ
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Удалить
    </button>
  </div>
  <div>
    <label>
      <input
        checked={false}
        name="checkbox_3"
        onChange={[Function]}
        type="checkbox"
      />
      Приготовить ужин
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Удалить
    </button>
  </div>
  <div>
    <label>
      <input
        checked={false}
        name="checkbox_4"
        onChange={[Function]}
        type="checkbox"
      />
      Уборка квартиры
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Удалить
    </button>
  </div>
  <div>
    <label>
      <input
        checked={true}
        name="checkbox_0"
        onChange={[Function]}
        type="checkbox"
      />
      Полить цветы
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Удалить
    </button>
  </div>
  <div>
    <label>
      <input
        checked={true}
        name="checkbox_2"
        onChange={[Function]}
        type="checkbox"
      />
      Купить продукты
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Удалить
    </button>
  </div>
</div>
`;
